odoo/modules/registry.py

```diff
@@ -495,7 +495,7 @@ def check_foreign_keys(self, cr):
             if spec is None:
                 sql.add_foreign_key(cr, table1, column1, table2, column2, ondelete)
                 model.env['ir.model.constraint']._reflect_constraint(model, conname, 'f', None, module)
-            elif spec != (conname, table2, column2, deltype):
+            elif (spec[1], spec[2], spec[3]) != (table2, column2, deltype):
                 sql.drop_constraint(cr, table1, spec[0])
                 sql.add_foreign_key(cr, table1, column1, table2, column2, ondelete)
                 model.env['ir.model.constraint']._reflect_constraint(model, conname, 'f', None, module)
```

Comment:
Actually, the `conname` stored in `ir.model.constraint` will be wrong. 
Actually, I think that's the computation of `conname` that should be fixed instead.

Reply:
I agree with @KangOl : `conname` is indeed the source of the problem.
I actually wonder whether it is safe to ignore the constraint name in this context...